andersoncd/solver.py

```diff
@@ -246,21 +243,23 @@ def solver(
 
 
 @njit
-def _kkt_violation(w, X, R, penalty, ws):
+def _kkt_violation(w, X, Xw, datafit, penalty, ws):
     n_samples = X.shape[0]
     grad = np.zeros(ws.shape[0])
     for idx, j in enumerate(ws):
-        grad[idx] = X[:, j] @ R / n_samples
+        grad[idx] = datafit.gradient_scalar(X, w, Xw, j) / n_samples  # TODO
```

Comment:
I would remove the n_samples and implement the exact gradient, less error prone for the future IMO